services/s3/generated.go

```diff
@@ -82,269 +73,288 @@ func GetStorageSystemMetadata(s *StorageMeta) StorageSystemMetadata {
-			}
-			result.HasCredential = true
-			result.Credential = v.Value.(string)
-		case "default_service_pairs":
-			if result.HasDefaultServicePairs {
-				continue
-			}
-			result.HasDefaultServicePairs = true
-			result.DefaultServicePairs = v.Value.(DefaultServicePairs)
+			f.Credential = v.Value.(string)
+		case "default_storage_class":
+			f.DefaultStorageClass = v.Value.(string)
+		case "enable_virtual_dir":
+			f.EnableVirtualDir = v.Value.(bool)
+		case "enable_virtual_link":
+			f.EnableVirtualLink = v.Value.(bool)
 		case "endpoint":
-			if result.HasEndpoint {
-				continue
-			}
-			result.HasEndpoint = true
-			result.Endpoint = v.Value.(string)
+			f.Endpoint = v.Value.(string)
 		case "force_path_style":
-			if result.HasForcePathStyle {
-				continue
-			}
-			result.HasForcePathStyle = true
-			result.ForcePathStyle = v.Value.(bool)
-		case "http_client_options":
-			if result.HasHTTPClientOptions {
-				continue
-			}
-			result.HasHTTPClientOptions = true
-			result.HTTPClientOptions = v.Value.(*httpclient.Options)
-		case "service_features":
-			if result.HasServiceFeatures {
-				continue
-			}
-			result.HasServiceFeatures = true
-			result.ServiceFeatures = v.Value.(ServiceFeatures)
+			f.ForcePathStyle = v.Value.(bool)
+		case "location":
+			f.Location = v.Value.(string)
+		case "name":
+			f.Name = v.Value.(string)
 		case "use_accelerate":
-			if result.HasUseAccelerate {
-				continue
-			}
-			result.HasUseAccelerate = true
-			result.UseAccelerate = v.Value.(bool)
+			f.UseAccelerate = v.Value.(bool)
 		case "use_arn_region":
-			if result.HasUseArnRegion {
-				continue
-			}
-			result.HasUseArnRegion = true
-			result.UseArnRegion = v.Value.(bool)
+			f.UseArnRegion = v.Value.(bool)
+		case "work_dir":
+			f.WorkDir = v.Value.(string)
 		}
 	}
-	// Enable features
-
-	// Default pairs
-
-	if !result.HasCredential {
-		return pairServiceNew{}, services.PairRequiredError{Keys: []string{"credential"}}
+	return nil
+}
+func (f *Factory) FromMap(m map[string]interface{}) (err error) {
+	return errors.New("FromMap not implemented")
+}
+func (f *Factory) NewServicer() (srv types.Servicer, err error) {
+	return f.newService()
+}
+func (f *Factory) NewStorager() (sto types.Storager, err error) {
+	return f.newStorage()
+}
+func (f *Factory) serviceFeatures() (s types.ServiceFeatures) {
+	s.Create = true
+	s.Delete = true
+	s.Get = true
+	s.List = true
+	return
+}
+func (f *Factory) storageFeatures() (s types.StorageFeatures) {
+	s.CompleteMultipart = true
+	s.Create = true
+	s.CreateDir = true
+	s.CreateLink = true
+	s.CreateMultipart = true
+	s.Delete = true
+	s.List = true
+	s.ListMultipart = true
+	s.Metadata = true
+	s.QuerySignHTTPRead = true
+	s.QuerySignHTTPWrite = true
+	s.QuerySignHTTPWriteMultipart = true
+	s.Read = true
+	s.Stat = true
+	s.Write = true
+	s.WriteMultipart = true
+	if f.EnableVirtualDir {
+		s.VirtualDir = true
 	}
-	return result, nil
+	if f.EnableVirtualLink {
+		s.VirtualLink = true
+	}
```

Comment:
Actually `write_empty_obejct` is also allowed. But I think maybe it should not be initialized with `WithEnableWriteEmptyObject` as it's the service capability?

Reply:
Nice catch, I will find a way to distinguish them.

Reply:
Fixed in https://github.com/beyondstorage/go-storage/pull/1008